runtime/intern.c

```diff
@@ -776,16 +799,16 @@ static value caml_input_val_core(struct channel *chan, int outside_heap)
   intern_alloc(h.whsize, h.num_objects, outside_heap);
   /* Fill it in */
   intern_rec(&res);
-  if (!outside_heap) {
-    intern_add_to_heap(h.whsize);
-  } else {
+  if (!outside_heap)
+    return intern_end(res, h.whsize);
+  else {
     caml_disown_for_heap(intern_extra_block);
     intern_extra_block = NULL;
     intern_block = 0;
+    /* Free everything */
+    intern_cleanup();
+    return caml_check_urgent_gc(res);
```

Comment:
Note: I think it would be nicer to have the same level of abstraction in the success and failure cases, with two `intern_end_{success,failure}` functions .